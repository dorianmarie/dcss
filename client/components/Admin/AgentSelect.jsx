import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Dropdown, Table } from '@components/UI';

const AgentSelectItem = ({ title, description }) => {
  return (
    <Table>
      <Table.Header>
        <Table.Row>
          <Table.HeaderCell>{title}</Table.HeaderCell>
        </Table.Row>
      </Table.Header>
      <Table.Body>
        {description ? (
          <Table.Row>
            <Table.Cell>{description}</Table.Cell>
          </Table.Row>
        ) : null}
      </Table.Body>
    </Table>
  );
};

AgentSelectItem.propTypes = {
  description: PropTypes.string,
  title: PropTypes.string
};

class AgentSelect extends Component {
  constructor(props) {
    super(props);
    this.onChange = this.onChange.bind(this);
  }

  onChange(e, { value }) {
    const selected = value ? this.props.agentsById[value] : null;
    this.props.onSelect(selected);
  }

  render() {
    const { onChange } = this;
    const {
      emptyText = 'No selection',
      fluid,
      item,
      agents,
      placeholder,
      search,
      selection,
      types,
      value,
      error
    } = this.props;

    // This empty "agent" option is used to prevent the
    // dropdown from treating "open" as "select".
    agents.unshift({
      id: '',
      description: '',
      title: emptyText
    });

    const options = agents.reduce((accum, agent) => {
      const option = {
        key: agent.id,
        value: agent.id,
        text: agent.title
      };

      if (agent.id && agent.interaction) {
        option.content = <AgentSelectItem {...agent} />;

        const isSupported =
          agent.interaction.types &&
          types.every(type => agent.interaction.types.includes(type));
        if (isSupported) {
          // Only display agents that are compatible
          // with this component type
          accum.push(option);
        }
      } else {
        // Include the dummy.
        accum.push(option);
      }
      return accum;
    }, []);

    const scrolling = true;
    const dropdownProps = {
      'aria-label': placeholder,
      fluid,
      item,
      onChange,
      options,
      placeholder,
      scrolling,
      search,
      selection,
      value,
      error
    };
    return <Dropdown {...dropdownProps} />;
  }
}

AgentSelect.propTypes = {
  agents: PropTypes.array,
  agentsById: PropTypes.object,
  agentsInUse: PropTypes.array,
  defaultValue: PropTypes.number,
  error: PropTypes.bool,
  emptyText: PropTypes.string,
  fluid: PropTypes.bool,
  item: PropTypes.any,
  onSelect: PropTypes.func,
  placeholder: PropTypes.string,
  search: PropTypes.bool,
  selection: PropTypes.bool,
  style: PropTypes.object,
  types: PropTypes.array,
  value: PropTypes.number
};

const mapStateToProps = (state, ownProps) => {
  const types = ownProps.types && ownProps.types.length ? ownProps.types : [];
  const agentsInUse = ownProps.agentsInUse;
  const sourceAgents = state.agents.length
    ? state.agents
    : ownProps.agents || [];

  // Ensure that we only display agents that are actually in use.
  // If none are in use, then the component does nothing.
  const preFilteredAgents = agentsInUse
    ? sourceAgents.filter(a => agentsInUse.includes(a.id))
    : sourceAgents;

  const filteredAgents = preFilteredAgents.filter(a => !a.deleted_at);

  const emptyText = !filteredAgents.length
    ? 'There are currently no compatible agents in use'
    : ownProps.emptyText;

  const placeholder = !filteredAgents.length ? emptyText : ownProps.placeholder;

  const agents = filteredAgents.slice();
  const agentsById = agents.reduce(
    (accum, agent) => ({
      ...accum,
      [agent.id]: agent
    }),
    {}
  );

  return {
    agents,
    agentsInUse,
    agentsById,
    emptyText,
    placeholder,
    types
  };
};

const mapDispatchToProps = dispatch => ({
  // getAgents: () => dispatch(getAgents())
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AgentSelect);
