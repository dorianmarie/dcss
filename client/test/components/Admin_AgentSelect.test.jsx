/** @TEMPLATE: BEGIN **/
import React from 'react';
jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useLayoutEffect: jest.requireActual('react').useEffect,
}));

import {
  fetchImplementation,
  mounter,
  reduxer,
  serialize,
  snapshotter,
  state,
} from '../bootstrap';
import { unmountComponentAtNode } from 'react-dom';

import {
  act,
  fireEvent,
  prettyDOM,
  render,
  screen,
  waitFor,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';

async function waitForPopper() {
  // Popper update() - https://github.com/popperjs/react-popper/issues/350
  await act(async () => await null);
}

/** @TEMPLATE: END **/

/** @GENERATED: BEGIN **/

let agents;
let agentsById;

import AgentSelect from '../../components/Admin/AgentSelect.jsx';
/** @GENERATED: END **/

/** @TEMPLATE: BEGIN **/
const original = JSON.parse(JSON.stringify(state));
let container = null;
let commonProps = null;
let commonState = null;
/** @TEMPLATE: END **/

beforeAll(() => {
  /** @TEMPLATE: BEGIN **/
  (window || global).fetch = jest.fn();
  /** @TEMPLATE: END **/
});

afterAll(() => {
  /** @TEMPLATE: BEGIN **/
  jest.restoreAllMocks();
  /** @TEMPLATE: END **/
});

beforeEach(() => {
  /** @TEMPLATE: BEGIN **/
  jest.useFakeTimers();
  container = document.createElement('div');
  container.setAttribute('id', 'root');
  document.body.appendChild(container);

  fetchImplementation(fetch);
  /** @TEMPLATE: END **/

  /** @GENERATED: BEGIN **/

  agent = {
    id: 1,
    created_at: '2021-02-25T17:31:33.826Z',
    updated_at: '2021-02-25T20:09:04.999Z',
    deleted_at: null,
    is_active: true,
    title: 'Emoji Analysis',
    name: 'emoji-analysis',
    description: 'Detects the presense of an emoji character in your text',
    endpoint: 'ws://emoji-analysis-production.herokuapp.com',
    configuration: {
      bar: '2',
      baz: 'c',
      foo: 'false',
    },
    interaction: {
      id: 1,
      name: 'ChatPrompt',
      description:
        'It will appear as an option for scenario authors to include in chat discussions within multi-participant scenarios. It receives participant chat messages.',
      created_at: '2021-02-25T15:09:05.001302-05:00',
      deleted_at: null,
      updated_at: null,
      types: [],
    },
    owner: {
      id: 999,
      email: 'super@email.com',
      roles: ['participant', 'super_admin', 'facilitator', 'researcher'],
      is_super: true,
      username: 'superuser',
      is_anonymous: false,
      personalname: 'Super User',
    },
    self: {
      id: 148,
      email: null,
      roles: null,
      is_super: false,
      username: 'ebe565050b31cbb4e7eacc39b23e2167',
      lastseen_at: '2021-02-25T13:08:57.323-05:00',
      is_anonymous: true,
      personalname: 'Emoji Analysis',
      single_use_password: false,
    },
    socket: {
      path: '/path/to/foo',
    },
  };

  agents = [agent];

  agentsById = agents.reduce((accum, agent) => {
    accum[agent.id] = agent;
    return accum;
  }, {});

  /** @GENERATED: END **/

  /** @TEMPLATE: BEGIN **/
  commonProps = {};
  commonState = JSON.parse(JSON.stringify(original));
  /** @TEMPLATE: END **/
});

afterEach(() => {
  /** @TEMPLATE: BEGIN **/
  jest.runOnlyPendingTimers();
  jest.useRealTimers();
  jest.resetAllMocks();
  unmountComponentAtNode(container);
  container.remove();
  container = null;
  commonProps = null;
  commonState = null;
  /** @TEMPLATE: END **/
});

test('AgentSelect', () => {
  expect(AgentSelect).toBeDefined();
});

/** @GENERATED: BEGIN **/
test('Render 1 1', async (done) => {
  const Component = AgentSelect;
  const props = {
    ...commonProps,
    agents,
    onSelect: jest.fn(),
  };

  const state = {
    ...commonState,
  };

  const ConnectedRoutedComponent = reduxer(Component, props, state);

  await render(<ConnectedRoutedComponent {...props} />);
  expect(serialize()).toMatchSnapshot();

  done();
});
/** @GENERATED: END **/

