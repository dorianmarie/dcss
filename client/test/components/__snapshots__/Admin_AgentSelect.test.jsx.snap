// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render 1 1 1`] = `
"<body>
  <div
    id=\\"root\\"
  />
  <div />
</body>"
`;
